libcob/common.h

```diff
@@ -45,7 +45,17 @@ typedef __mpz_struct mpz_t[1];
 #define	cob_sli_t		long int
 #define	cob_uli_t		unsigned long int
 
-#if	defined(_WIN32) && !defined(__MINGW32__)
+#ifdef HAVE_STDINT_H
+
+#include <stdint.h>
+
+#define	cob_s64_t		int64_t
+#define	cob_u64_t		uint64_t
+
+#define	COB_S64_C(x)		x ## LL
+#define	COB_U64_C(x)		x ## ULL
+
+#elif	defined(_WIN32) && !defined(__MINGW32__)
```

Comment:
We can drop the MINGW part here, even over 10 year old MinGW ships a header from 2000-12-02 which has those types.

Reply:
Out of curiosity, do you see any reason why one would want to compile GnuCobol with any C compiler that is not at least C99-compliant ? (besides people using outdated MSVC versions...)

Reply:
Legacy MSVC was one of the reasons years ago, but as this finally improved since years (somewhere 2013/2015, I think) this isn't much of a reason any more.
The likely oldest environment would be [GCCMVS](https://gccmvs.sourceforge.net/) - That's GCC 3.2.3 as 31bit compiler (but I never tried compiling any GnuCOBOL version using that compiler).

COBOL tackles quite some legacy systems. And it is easier to first only move out a proprietary compiler (which may have license issues) to GnuCOBOL (so all your files are where expected scripts work as before, ...) than to move out everything.

My direct main comparison to ACUCOBOL for example is a SUSE 6.4 VM (which ships its compiler and runtime :-) and that is GCC but not C99. [more a hobby machine, but there are likely people out there which have other ancient environments running their COBOL].

We _may_ require C99 with GC 4.0+. The more important "legacy" systems I'm aware of like AIX and Solaris don't do C99 by default, but you can tweak this with the appropriate options (then specified together with the compiler as `CC` during configure), but even there we have no clear date; which is also one of the reasons that we have a CI using an old std for GCC (which doesn't set limits on all parts, so that's a "help", not a "guarantee").

Reply:
As-is GnuCOBOL expects `int` to be an integer type > 16bit and expects an integer type > 32 bit to be available (both since at least 3.x, possibly also since 2.2) .

`long long` is very portable (outside of Windows where its own type may be used).
Your concern is that the newly used headers are available, but miss `long long` as definition for a 64bit type, right?

In this case the new code could have something like `#ifdef ADD_COB64` then placing the previously used `long long` definition in.

OT: I've learned yesterday that Hercules won't work as cross-compile, as there is no full cross-compiling environment available (you can cross-compile, but this leads to ELF files, so you would only generate to C then copy over, then assemble with the machine native assembler).

Reply:
> Your concern is that the newly used headers are available, but miss long long as definition for a 64bit type, right?

My concern is that `stdint.h` and `inttypes.h` always assume the smallest possible type for 64-bit integers, so it will use `long` on 64-bit Linux but `long long` on 32-bit Linux and 32/64-bit Windows with MinGW (don't know about MSVC).
However GnuCobol takes the approach of always using `long long` (ignoring MSVC), which is safe and avoid the trouble of having to choose the correct type. However even when `sizeof(long)` = `sizeof(long long)`, the two are considered different types and may generate warnings if one is use in place of the other (this also affects the `L` and `LL` suffixes). That's what I'm trying to avoid.

In the first place `stdint.h` and `inttypes.h` got involved because I wanted to use the `PRId64` macro to avoid having to discriminate between the standard `lld` and MSVCRT-specific `I64d` format specifiers.

Maybe we can do that : use `int64_t` and `PRId64` from `stdint.h` and `inttypes.h` where available (and in particular to deal with the mess about MSVC), and just fallback to plain `lond long` and `lld` otherwise (safe fallback for all those C89 compilers that do support `long long` as an extension). This would break compilation on old versions of Visual C++ (< 2010 or something) (although we could use MSC_VER to add a fallback for these versions) ; not sure about MinGW (though the oldest versions I could find does have `stdint.h` and `inttypes.h`).